src/Akka.Cluster.Hosting/AkkaClusterHostingExtensions.cs

```diff
@@ -36,6 +38,17 @@ public sealed class ClusterSingletonOptions
         public int? BufferSize { get; set; } = null;
         public string Role { get; set; }
         public object TerminationMessage { get; set; }
+        
+        /// <summary>
+        /// An class instance that extends <see cref="LeaseOptionBase"/>, used to configure the lease provider used in this
+        /// cluster singleton.
+        /// </summary>
+        public LeaseOptionBase? LeaseImplementation { get; set; }
```

Comment:
I think this is kind of ugly - reason being is that all of the other `*Options` classes are POCOs but `LeaseOptionsBase` combines effects (`Apply(AkkaConfigurationBuilder builder, Setup setup = null);`) and data - it'd be cleaner if we could keep the implementations pure. How could we go about doing this?

Reply:
`LeaseOptionBase` is the exact same way we're doing it in SBR, the same identical option can be used for both SBR and sharding at the same time.

Reply:
That part I get, but it's the combination of state + effects in one of the configuration classes that I'm iffy on.

Reply:
It is practically (for now) a placeholder to point the sharding plugin to the actual configuration path being used by the implementing `Lease` class itself.
For example, the implementation in kubernetes is here: https://github.com/akkadotnet/Akka.Management/blob/51638f2acb51efaea64415fc5dd7df214e49f63b/src/coordination/kubernetes/Akka.Coordination.KubernetesApi/KubernetesLeaseOption.cs

Reply:
It does have the ability to modify/inject more code into `AkkaConfigurationBuilder` if needed, that part is a future proofing feature. `IHoconOption` itself is __not__ intended for end user, it is more of a template an `Akka.Hosting` developer had to use to implement a very specific HOCON pattern. The documentation is here: https://github.com/akkadotnet/Akka.Hosting/blob/6502ac3ac58490b2514befd4877e6b69a20aeb93/src/Akka.Hosting/IHoconOption.cs#L12-L62

Reply:
Ok, we'll leave that be then

Reply:
Prettyfying it:

# IHoconOption

Standardized interface template for a common HOCON configuration pattern where a configuration takes a HOCON config path and the config path contains a class FQCN property and other settings that are needed by said class.

The pattern looks like this:
```hocon
# This HOCON property references to a config block below
akka.discovery.method = akka.discovery.config

akka.discovery.config {
    class = "Akka.Discovery.Config.ConfigServiceDiscovery, Akka.Discovery"
    # other options goes here
}
```
## Example
Example implementation for the pattern described above
```csharp
// The base class for the option (templating)
public abstract class DiscoveryOptionBase : IOption
{ }

// The actual option implementation
public class ConfigOption : DiscoveryOptionBase
{
    // Actual option implementation here
     public void Apply(AkkaConfigurationBuilder builder)
     {
         // Modifies Akka.NET configuration either via HOCON or setup class
         builder.AddHocon($"akka.discovery.method = {ConfigPath}", HoconAddMode.Prepend);

         // Rest of configuration goes here
     }
}

 // Akka.Hosting extension implementation
 public static AkkaConfigurationBuilder WithDiscovery(
     this AkkaConfigurationBuilder builder,
     DiscoveryOptionBase discOption)
 {
     var setup = new DiscoverySetup();

     // gets called here
     discOption.Apply(builder, setup);
 }
```

Reply:
The user isn't supposed to see all this horribleness, what they needed to do is to set
`new ClusterSingletonOptions { LeaseImplementation = KubernetesLeaseOption.Instance };`